pennylane/transforms/adjoint_metric_tensor.py

```diff
@@ -0,0 +1,322 @@
+    """Wrapper that allows to apply a variety of operations---or groups
+    of operations---to a state or to prepare a new state.
+    If ``invert=True``, this function makes sure not to alter the operations.
+    The state of the device, however may be altered, depending on the
+    device and performed operation(s).
+    """
+    # pylint: disable=protected-access
+    if isinstance(op, (list, np.ndarray)):
+        if invert:
+            op = op[::-1]
+        for _op in op:
+            state = _apply_operations(state, _op, device, invert)
+        return state
+
+    if isinstance(op, qml.QubitStateVector):
+        if invert:
+            raise ValueError("Can't invert state preparation.")
+        device._apply_state_vector(op.parameters[0], op.wires)
+        return device._state
+
+    if isinstance(op, qml.BasisState):
+        if invert:
+            raise ValueError("Can't invert state preparation.")
+        device._apply_basis_state(op.parameters[0], op.wires)
+        return device._state
+
+    if invert:
+        op.inv()
+    state = device._apply_operation(state, op)
+    if invert:
+        op.inv()
+
+    return state
+
+
+def _group_operations(tape):
+    """Divide all operations of a tape into trainable operations and blocks
+    of untrainable operations after each trainable one."""
+
+    # Extract tape operations list
+    ops = tape.operations
+    # Find the indices of trainable operations in the tape operations list
+    trainables = np.where([qml.operation.is_trainable(op) for op in ops])[0]
+    # Add the indices incremented by one to the trainable indices
+    split_ids = list(chain.from_iterable([idx, idx + 1] for idx in trainables))
+
+    # Split at trainable and incremented indices to get groups after trainable
+    # operations and single trainable operations (in alternating order)
+    all_groups = np.split(ops, split_ids)
+
+    # Collect trainable operations and groups after trainable operations
+    # the first set of non-trainable ops are the ops "after the -1st" trainable op
+    group_after_trainable_op = dict(enumerate(all_groups[::2], start=-1))
+    trainable_operations = list(chain.from_iterable(all_groups[1::2]))
+
+    return trainable_operations, group_after_trainable_op
+
+
+def adjoint_metric_tensor(circuit, device=None, hybrid=True):
+    r"""Implements the adjoint method outlined in
+    `Jones <https://arxiv.org/abs/2011.02991>`__ to compute the metric tensor.
+
+    A forward pass followed by intermediate partial backwards passes are
+    used to evaluate the metric tensor in :math:`\mathcal{O}(p^2)` operations,
+    where :math:`p` is the number of trainable operations, using 4 state
+    vectors.
+
+    .. note::
+        The adjoint metric tensor method has the following restrictions:
+
+        * As it requires knowledge of the statevector and uses private
+          methods of the used device, only ``"default.qubit"`` devices can be used.
```

Comment:
```suggestion
        * Currently only ``"default.qubit"`` with ``shots=None`` is supported.
```

Might be better to make this much more concise/explicit! It is easy to miss otherwise